cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs

```diff
@@ -0,0 +1,741 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
+module Cardano.CLI.Shelley.Run.Read
+  ( -- * Metadata
+    MetadataError(..)
+  , renderMetadataError
+  , readFileTxMetadata
+  , readTxMetadata
+
+    -- * Script
+  , ScriptWitnessError(..)
+  , renderScriptWitnessError
+  , readScriptDataOrFile
+  , readScriptWitness
+  , readScriptWitnessFiles
+  , readScriptWitnessFilesThruple
+  , ScriptDecodeError (..)
+  , deserialiseScriptInAnyLang
+  , readFileScriptInAnyLang
+
+  -- * Script data (datums and redeemers)
+  , ScriptDataError(..)
+  , readScriptDatumOrFile
+  , readScriptRedeemerOrFile
+  , renderScriptDataError
+
+  -- * Tx
+  , CddlError
+  , CddlTx(..)
+  , IncompleteTx(..)
+  , readFileTx
+  , readFileTxBody
+  , renderCddlError
+  , readCddlTx -- For testing purposes
+
+  -- * Tx witnesses
+  , ReadWitnessSigningDataError(..)
+  , renderReadWitnessSigningDataError
+  , SomeWitness(..)
+  , ByronOrShelleyWitness(..)
+  , ShelleyBootstrapWitnessSigningKeyData(..)
+  , CddlWitnessError(..)
+  , readFileTxKeyWitness
+  , readWitnessSigningData
+
+  -- * Required signer
+  , RequiredSignerError(..)
+  , categoriseSomeWitness
+  , readRequiredSigner
+  ) where
+
+import           Prelude
+
+import           Control.Monad.Trans.Except (ExceptT (..), runExceptT)
+import           Control.Monad.Trans.Except.Extra (firstExceptT, handleIOExceptT, hoistEither, left)
+import qualified Data.Aeson as Aeson
+import           Data.Bifunctor (first)
+import qualified Data.ByteString.Char8 as BS
+import qualified Data.ByteString.Lazy.Char8 as LBS
+import qualified Data.List as List
+import qualified Data.Text as Text
+import           Data.Word
+
+
+import           Cardano.Api
+import           Cardano.Api.Shelley
+
+--TODO: do this nicely via the API too:
+import qualified Cardano.Binary as CBOR
+import           Data.Text (Text)
+--TODO: following import needed for orphan Eq Script instance
+import           Cardano.Ledger.Shelley.Scripts ()
+
+import           Cardano.CLI.Shelley.Key
+import           Cardano.CLI.Shelley.Parsers
+import           Cardano.CLI.Types
+
+-- Metadata
+
+data MetadataError
+  = MetadataErrorFile (FileError ())
+  | MetadataErrorJsonParseError !FilePath !String
+  | MetadataErrorConversionError !FilePath !TxMetadataJsonError
+  | MetadataErrorValidationError !FilePath ![(Word64, TxMetadataRangeError)]
+  | MetadataErrorDecodeError !FilePath !CBOR.DecoderError
+  | MetadataErrorNotAvailableInEra AnyCardanoEra
+
+renderMetadataError :: MetadataError -> Text
+renderMetadataError (MetadataErrorFile fileErr) =
+  Text.pack $ displayError fileErr
+renderMetadataError (MetadataErrorJsonParseError fp jsonErr) =
+  Text.pack $ "Invalid JSON format in file: " <> show fp <>
+              "\nJSON parse error: " <> jsonErr
+renderMetadataError (MetadataErrorConversionError fp metadataErr) =
+  Text.pack $ "Error reading metadata at: " <> show fp <>
```

Comment:
Sometimes we say "in file: <fp>" and sometimes we say "at: <fp>".

Is there something different going on?  Or should we use the same language in all cases?

Reply:
Nope just an inconsistency. I wouldn't fix that in this PR however. Open an issue if you think its worth fixing.